### ISSUE TITLE
Generate mage_output_file.go outside workdir

### ISSUE THREAD

**Issue by goenning** [magefile/mage#182]
So I've run into the situation where I use `mage run` to execute my app, while having `https://github.com/cosmtrek/air` watching the workdir looking for any file change. If there's a file change, then air rerun the app using `mage run`. The thing is that every time `mage run` is executed, a `mage_output_file.go` is generated on the workdir. 

1) Can we make it generate inside a tmp folder? Or even inside the `$HOME/.magefile`

Basically change this `main := filepath.Join(inv.Dir, mainfile)` to `main := filepath.Join(inv.CacheDir, mainfile)`

2) Even better, can we not generate a `mage_output_file.go` if makefile.go hasn't changed?

**Comment by natefinch** [comment]
So, this is tricky.   Building code outside of gopath is problematic (or outside a module in module-world). 

What you can probably do is use the `-compile` flag to create a binary in the current directory, and then tell air to run that.  

Something like this:

```go
func Air() error {
    if err := sh.Run("mage", "-compile=runner"); err != nil {
        return err
    }
    if err := sh.Run("air"); err != nil {
        return err
    } 
}

func Run() error {
    // your normal running code
    return nil
}
```

And then configure air with 

```
[build]
cmd = "./runner run"
```

Then you'd start it up by running `mage air`

**Comment by natefinch** [comment]
Oh, I had an answer to that all typed out but then didn't send it.  That's actually what mage used to do, but the problem is that we can't tell what dependencies might have changed, so people would change a library their magefile used and then mage would think nothing changed and wouldn't rebuild.

You can actually still get that behavior if you run mage with an older version of go (anything before 1.10).  

As of mage 1.6, we always run go build, but it will only rebuild if something has changed. It only takes about 100ms more than it did when we I was just checking the hashes of files, so I think it's a good tradeoff to have a much more accurate account of if anything has changed.

I can look at building in a subdirectory, but I am afraid that might have unintended side effects.   I think it would be better if air could just ignore a file.  That seems like a useful feature for it to have anyway.

**Comment by goenning** [comment]
Alright, that makes sense, didn’t think about that 😅

So unless there’s another reason, I wouldn’t build it inside a subfolder. I’ll send a PR to air. Feel free to close this one!

Thanks.
